### PR TITLE
WIP: Add Bazel build files

### PR DESCRIPTION
**Pull request by munnerz**
**What this PR does / why we need it**:

This is an experiment to add support for building Navigator with Bazel. It should eventually replace `make`, and will help reduce the number of dependencies we need to bake into our test-infra images.

This PR also allows building docker images without Docker itself. It can currently also be used to build and unit test (i.e. gotest) all components of Navigator.

**Special notes for your reviewer**:

This is WIP, and work needs to be done to make this work in our CI environment as well as triggering e2e tests (and other scripts in the `hack/` directory)

**Release note**:
```release-note
NONE
```

### CODE REVIEW

**Comment by munnerz** [comment]
Bazel verify job is passing :tada:

Now just to move more things into this job!

**Review by wallrj — COMMENTED** [review]
```
 richard   munnerz-bazel  …  github.com  jetstack  navigator  130  make test
bazel test \
	--features=race \
	//cmd/... \
	//pkg/... \
	//internal/... \
	//plugin/...
Extracting Bazel installation...
................
INFO: Analysed 107 targets (548 packages loaded).
INFO: Found 84 targets and 23 test targets...
ERROR: missing input file '//pkg/controllers/generic:generic.go'
ERROR: /home/richard/projects/golang/src/github.com/jetstack/navigator/pkg/controllers/generic/BUILD.bazel:3:1: //pkg/controllers/generic:go_default_library: missing input file '//pkg/controllers/generic:generic.go'
ERROR: /home/richard/projects/golang/src/github.com/jetstack/navigator/pkg/controllers/generic/BUILD.bazel:3:1 1 input file(s) do not exist
INFO: Elapsed time: 48.177s, Critical Path: 7.44s
FAILED: Build did NOT complete successfully
//internal/test/unit/framework:go_default_xtest                       NO STATUS
//internal/test/util:go_default_test                                  NO STATUS
//pkg/apis/navigator/validation:go_default_xtest                      NO STATUS
//pkg/cassandra/nodetool:go_default_xtest                             NO STATUS
//pkg/cassandra/version:go_default_xtest                              NO STATUS
//pkg/controllers/cassandra:go_default_xtest                          NO STATUS
//pkg/controllers/cassandra/nodepool:go_default_xtest                 NO STATUS
//pkg/controllers/cassandra/pilot:go_default_xtest                    NO STATUS
//pkg/controllers/cassandra/role:go_default_xtest                     NO STATUS
//pkg/controllers/cassandra/rolebinding:go_default_xtest              NO STATUS
//pkg/controllers/cassandra/seedlabeller:go_default_xtest             NO STATUS
//pkg/controllers/cassandra/service:go_default_xtest                  NO STATUS
//pkg/controllers/cassandra/serviceaccount:go_default_xtest           NO STATUS
//pkg/controllers/elasticsearch/actions:go_default_test               NO STATUS
//pkg/controllers/elasticsearch/role:go_default_test                  NO STATUS
//pkg/controllers/elasticsearch/rolebinding:go_default_test           NO STATUS
//pkg/pilot/cassandra/v3:go_default_test                              NO STATUS
//pkg/pilot/elasticsearch/v5:go_default_test                          NO STATUS
//pkg/pilot/genericpilot/controller/scheduler:go_default_test         NO STATUS
//pkg/util:go_default_test                                            NO STATUS
//pkg/util/api:go_default_test                                        NO STATUS
//pkg/util/hash:go_default_test                                       NO STATUS
//plugin/pkg/admission/namespace/lifecycle:go_default_test            NO STATUS

Executed 0 out of 23 tests: 23 were skipped.
make: *** [Makefile:100: go_test] Error 1
```

File: hack/update-bazel.sh [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+
+set -o errexit
+set -o nounset
+set -o pipefail
+
+SCRIPT_ROOT=$(dirname ${BASH_SOURCE})/..
+
+echo "Running 'bazel run //:gazelle'"
+
+bazel run //:gazelle
```
[the remark this answers is not in the document]
Comment (wallrj) [reply hack/update-bazel.sh]: And when I run it it adds an additional bazel file.
```
diff --git a/hack/update-bazel.sh b/hack/update-bazel.sh
old mode 100644
new mode 100755
diff --git a/pkg/controllers/generic/BUILD.bazel b/pkg/controllers/generic/BUILD.bazel
index 2ef02aec..e69de29b 100644
--- a/pkg/controllers/generic/BUILD.bazel
+++ b/pkg/controllers/generic/BUILD.bazel
@@ -1,8 +0,0 @@
-load("@io_bazel_rules_go//go:def.bzl", "go_library")
-
-go_library(
-    name = "go_default_library",
-    srcs = ["generic.go"],
-    importpath = "github.com/jetstack/navigator/pkg/controllers/generic",
-    visibility = ["//visibility:public"],
-)
```

File: hack/update-client-gen.sh [hunk 2]
```diff
@@ -6,7 +6,7 @@ set -o errexit
 set -o nounset
 set -o pipefail
 
-SCRIPT_ROOT=$(dirname ${BASH_SOURCE})/..
+SCRIPT_ROOT=$(dirname $(realpath "${BASH_SOURCE}"))/..
```
Comment (wallrj) [inline hack/update-client-gen.sh]: Should we use `realpath` in the other scripts too....or is this one in particular a symlink sometimes?

File: hack/update-bazel.sh (hunk 1 above)
[the remark this answers is not in the document]
Comment (wallrj) [reply hack/update-bazel.sh]: And now the tests pass,
```
make test
bazel test \
	--features=race \
	//cmd/... \
	//pkg/... \
	//internal/... \
	//plugin/...
INFO: Analysed 106 targets (534 packages loaded).
INFO: Found 83 targets and 23 test targets...
INFO: Elapsed time: 151.309s, Critical Path: 63.54s
INFO: Build completed successfully, 868 total actions
//internal/test/unit/framework:go_default_xtest                          PASSED in 3.5s
//internal/test/util:go_default_test                                     PASSED in 1.1s
//pkg/apis/navigator/validation:go_default_xtest                         PASSED in 2.7s
//pkg/cassandra/nodetool:go_default_xtest                                PASSED in 1.5s
//pkg/cassandra/version:go_default_xtest                                 PASSED in 1.1s
//pkg/controllers/cassandra:go_default_xtest                             PASSED in 2.2s
//pkg/controllers/cassandra/nodepool:go_default_xtest                    PASSED in 2.3s
//pkg/controllers/cassandra/pilot:go_default_xtest                       PASSED in 2.9s
//pkg/controllers/cassandra/role:go_default_xtest                        PASSED in 2.4s
//pkg/controllers/cassandra/rolebinding:go_default_xtest                 PASSED in 2.0s
//pkg/controllers/cassandra/seedlabeller:go_default_xtest                PASSED in 3.8s
//pkg/controllers/cassandra/service:go_default_xtest                     PASSED in 3.6s
//pkg/controllers/cassandra/serviceaccount:go_default_xtest              PASSED in 2.4s
//pkg/controllers/elasticsearch/actions:go_default_test                  PASSED in 11.2s
//pkg/controllers/elasticsearch/role:go_default_test                     PASSED in 2.0s
//pkg/controllers/elasticsearch/rolebinding:go_default_test              PASSED in 1.9s
//pkg/pilot/cassandra/v3:go_default_test                                 PASSED in 2.3s
//pkg/pilot/elasticsearch/v5:go_default_test                             PASSED in 2.1s
//pkg/pilot/genericpilot/controller/scheduler:go_default_test            PASSED in 13.1s
//pkg/util:go_default_test                                               PASSED in 1.1s
//pkg/util/api:go_default_test                                           PASSED in 2.7s
//pkg/util/hash:go_default_test                                          PASSED in 1.3s
//plugin/pkg/admission/namespace/lifecycle:go_default_test               PASSED in 2.0s

Executed 23 out of 23 tests: 23 tests pass.
There were tests whose specified size is too big. Use the --test_verbose_timeout_warnings command line option to see which ones these are.

```

[the remark this answers is not in the document]
Comment (wallrj) [reply hack/update-bazel.sh]: But
```
make test --test_verbose_timeout_warnings
make: unrecognized option '--test_verbose_timeout_warnings'
```

**Review by wallrj — APPROVED** [review]
Thanks @munnerz 

Subsequent tests and builds do seem incredibly fast!

I broke a test and got the following output:

```
//pkg/util/hash:go_default_test                                 (cached) PASSED in 1.3s
//plugin/pkg/admission/namespace/lifecycle:go_default_test      (cached) PASSED in 2.0s
//pkg/controllers/cassandra/seedlabeller:go_default_xtest                FAILED in 2.9s
  /home/richard/.cache/bazel/_bazel_richard/c2cb7431d7cf967e2afc789b3b2f2b1d/execroot/__main__/bazel-out/k8-fastbuild/testlogs/pkg/controllers/cassandra/seedlabeller/go_default_xtest/test.log

Executed 1 out of 23 tests: 22 tests pass and 1 fails locally.
There were tests whose specified size is too big. Use the --test_verbose_timeout_warnings command line option to see which ones these are.
make: *** [Makefile:100: go_test] Error 3

```

It'd be useful to see a more verbose failure message on the console.

* I also think we should add some documentation in with this branch explaining how and when to run each of the new hack scripts and which version of bazel to install etc.

File: hack/print-workspace-status.sh [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+
+set -o errexit
+set -o nounset
+set -o pipefail
+
+SCRIPT_ROOT=$(dirname ${BASH_SOURCE})/..
+
+cat <<EOF
+STABLE_DOCKER_TAG build
+EOF
```
[the remark this answers is not in the document]
Comment (munnerz) [reply hack/print-workspace-status.sh]: This is used as the docker image tag used when producing docker image build artefacts.

This code needs updating to read the git tag from the environment. Thanks for reminding me 😄

**Pull request closed without merging**